In [7]:
import yfinance as yf
import pandas as pd

In [3]:
ticker = yf.Ticker("REL.L")
prices = ticker.history(period = "5y")
print(prices.head())
print(f"Number of rows: {len(prices)}")

                                  Open         High          Low        Close  \
Date                                                                            
2021-07-12 00:00:00+01:00  2000.976462  2034.942113  1998.978482  2026.950195   
2021-07-13 00:00:00+01:00  2031.944937  2061.914626  2019.957061  2056.919678   
2021-07-14 00:00:00+01:00  2044.931864  2079.135310  2027.949040  2047.928833   
2021-07-15 00:00:00+01:00  2040.936166  2065.910911  2030.946268  2033.943237   
2021-07-16 00:00:00+01:00  2047.928913  2058.917800  2036.940026  2053.922852   

                            Volume  Dividends  Stock Splits  
Date                                                         
2021-07-12 00:00:00+01:00  2706662        0.0           0.0  
2021-07-13 00:00:00+01:00  2951765        0.0           0.0  
2021-07-14 00:00:00+01:00  2787516        0.0           0.0  
2021-07-15 00:00:00+01:00  2333847        0.0           0.0  
2021-07-16 00:00:00+01:00  3113221        0.0           0.0 

In [4]:
info = ticker.info
print("Return on Equity:", info.get("returnOnEquity"))
print("Gross Margins:", info.get("grossMargins"))
print("Operating Margins:", info.get("operatingMargins"))
print("Debt to Equity:", info.get("debtToEquity"))
print("Revenue Growth:", info.get("revenueGrowth"))

Return on Equity: 0.70512
Gross Margins: 0.66288
Operating Margins: 0.31409
Debt to Equity: 304.059
Revenue Growth: 0.012


In [5]:
financials = ticker.financials
print(f"Financials: {financials}")

Financials:                                                       2025-12-31  \
Tax Effect Of Unusual Items                        -4.642909e+06   
Tax Rate For Calcs                                  2.443640e-01   
Normalized EBITDA                                   3.705000e+09   
Total Unusual Items                                -1.900000e+07   
Total Unusual Items Excluding Goodwill             -1.900000e+07   
Net Income From Continuing Operation Net Minori...  2.065000e+09   
Reconciled Depreciation                             6.630000e+08   
Reconciled Cost Of Revenue                          3.233000e+09   
EBITDA                                              3.686000e+09   
EBIT                                                3.023000e+09   
Net Interest Income                                -2.580000e+08   
Interest Expense                                    2.730000e+08   
Interest Income                                     1.200000e+07   
Normalized Income                   

In [6]:
balance_sheet = ticker.balance_sheet
print(f"Balance sheet: {balance_sheet}")

Balance sheet:                                                     2025-12-31    2024-12-31  \
Treasury Shares Number                            8.998919e+06  2.490282e+07   
Ordinary Shares Number                            1.819061e+09  1.855942e+09   
Share Issued                                      1.828060e+09  1.880845e+09   
Net Debt                                          7.039000e+09  6.322000e+09   
Total Debt                                        7.267000e+09  6.544000e+09   
...                                                        ...           ...   
Accounts Receivable                               2.162000e+09  2.184000e+09   
Allowance For Doubtful Accounts Receivable       -9.600000e+07 -1.220000e+08   
Gross Accounts Receivable                         2.258000e+09  2.306000e+09   
Cash Cash Equivalents And Short Term Investments  1.310000e+08  1.190000e+08   
Cash And Cash Equivalents                         1.310000e+08  1.190000e+08   

                        

In [8]:
sample_tickers = {
    "RELX": "REL.L",
    "Experian": "EXPN.L",
    "Sage": "SGE.L",
    "Unilever": "ULVR.L",
    "Diageo": "DGE.L",
    "Halma": "HLMA.L",
    "Intertek": "ITRK.L",
    "AstraZeneca": "AZN.L",
    "GSK": "GSK.L",
    "Bytes Technology": "BYIT.L",   
    "Games Workshop": "GAW.L",      
}

In [10]:
results = []

for name, ticker_symbol in sample_tickers.items():
    ticker = yf.Ticker(ticker_symbol)
    info = ticker.info

    prices = ticker.history(period = "5y")
    years_of_price_history = len(prices)

    has_roe = info.get("returnOnEquity") is not None

    financials = ticker.financials
    has_financials = len(financials) > 0
    years_of_financials = financials.shape[1] if has_financials else 0

    results.append({
        "Company": name,
        "Ticker": ticker_symbol,
        "Years of Price History": years_of_price_history,
        "Has Current ROE": has_roe,
        "Has Historical Financials": has_financials,
        "Years of Financials": years_of_financials,
    })

yfinance_data_audit = pd.DataFrame(results)
yfinance_data_audit

,Company,Ticker,Years of Price History,Has Current ROE,Has Historical Financials,Years of Financials
0,RELX,REL.L,1262,True,True,5
1,Experian,EXPN.L,1262,True,True,5
2,Sage,SGE.L,1262,True,True,5
3,Unilever,ULVR.L,1262,True,True,5
4,Diageo,DGE.L,1261,True,True,4
5,Halma,HLMA.L,1262,True,True,5
6,Intertek,ITRK.L,1262,True,True,5
7,AstraZeneca,AZN.L,1262,True,True,5
8,GSK,GSK.L,1262,True,True,5
9,Bytes Technology,BYIT.L,1261,True,True,5


## Data Audit Findings — yfinance

**Sample tested:** 11 FTSE 350 companies across sectors (info services, software, 
consumer staples, industrials, healthcare) and sizes (large-cap to small-cap).

**Results:**
- Price history: 11/11 companies had ~5 years of daily price data (~1,262 rows). 
  Consistent across large and small caps — no gaps found.
- Current fundamentals (ROE snapshot): 11/11 companies returned a value.
- Historical financials: 11/11 companies had historical financial statements 
  available, covering 4-5 years.

**Limitations found:**
1. **History depth is limited to ~5 years.** The project originally aimed to 
   backtest from ~2013, but yfinance's historical financials only go back 
   4-5 years. This may require either a paid data source for deeper history, 
   or narrowing the backtest period.
2. **No filing/report dates available.** yfinance only provides fiscal 
   period-end dates (e.g. "2023-12-31"), not the date results were actually 
   published. This creates a look-ahead bias risk — without filing dates, 
   we cannot be certain fundamentals were actually public knowledge at the 
   point they'd be used in the backtest. A fallback assumption (e.g. a fixed 
   ~90-120 day reporting lag) may be needed if a better data source doesn't 
   solve this.

**Conclusion:** yfinance is reliable and consistent for breadth (works across 
company size/sector), but insufficient on its own for depth and point-in-time 
accuracy. 

**Next step:** Test Financial Modeling Prep (FMP) on the same sample to 
see if it resolves either limitation.

In [11]:
from dotenv import load_dotenv
import os

In [ ]:
load_dotenv()
api_key = os.getenv("FMP_API_KEY")
print("Key loaded: ", api_key is not None)

Key loadded:  True


In [16]:
import requests

ticker_symbol = "AAPL"
url = f"https://financialmodelingprep.com/stable/income-statement?symbol={ticker_symbol}&apikey={api_key}"

response = requests.get(url)
data = response.json()

print(data)

[{'date': '2025-09-27', 'symbol': 'AAPL', 'reportedCurrency': 'USD', 'cik': '0000320193', 'filingDate': '2025-10-31', 'acceptedDate': '2025-10-31 06:01:26', 'fiscalYear': '2025', 'period': 'FY', 'revenue': 416161000000, 'costOfRevenue': 220960000000, 'grossProfit': 195201000000, 'researchAndDevelopmentExpenses': 34550000000, 'generalAndAdministrativeExpenses': 27601000000, 'sellingAndMarketingExpenses': 0, 'sellingGeneralAndAdministrativeExpenses': 27601000000, 'otherExpenses': 0, 'operatingExpenses': 62151000000, 'costAndExpenses': 283111000000, 'netInterestIncome': 0, 'interestIncome': 0, 'interestExpense': 0, 'depreciationAndAmortization': 11698000000, 'ebitda': 144427000000, 'ebit': 132729000000, 'nonOperatingIncomeExcludingInterest': 321000000, 'operatingIncome': 133050000000, 'totalOtherIncomeExpensesNet': -321000000, 'incomeBeforeTax': 132729000000, 'incomeTaxExpense': 20719000000, 'netIncomeFromContinuingOperations': 112010000000, 'netIncomeFromDiscontinuedOperations': 0, 'othe

In [17]:
print("Status code:", response.status_code)
print("Raw response text:", response.text)

Status code: 200
Raw response text: [
  {
    "date": "2025-09-27",
    "symbol": "AAPL",
    "reportedCurrency": "USD",
    "cik": "0000320193",
    "filingDate": "2025-10-31",
    "acceptedDate": "2025-10-31 06:01:26",
    "fiscalYear": "2025",
    "period": "FY",
    "revenue": 416161000000,
    "costOfRevenue": 220960000000,
    "grossProfit": 195201000000,
    "researchAndDevelopmentExpenses": 34550000000,
    "generalAndAdministrativeExpenses": 27601000000,
    "sellingAndMarketingExpenses": 0,
    "sellingGeneralAndAdministrativeExpenses": 27601000000,
    "otherExpenses": 0,
    "operatingExpenses": 62151000000,
    "costAndExpenses": 283111000000,
    "netInterestIncome": 0,
    "interestIncome": 0,
    "interestExpense": 0,
    "depreciationAndAmortization": 11698000000,
    "ebitda": 144427000000,
    "ebit": 132729000000,
    "nonOperatingIncomeExcludingInterest": 321000000,
    "operatingIncome": 133050000000,
    "totalOtherIncomeExpensesNet": -321000000,
    "incomeBefor

## Data Audit Findings — FMP (Financial Modeling Prep)

**Test:** Same income statement endpoint tested on RELX (REL.L, UK) vs AAPL (US, control test).

**Results:**
- REL.L (UK): HTTP 402 error — "Premium Query Parameter... not available under your current subscription." UK-listed tickers are not accessible on the free tier.
- AAPL (US): HTTP 200 success — full data returned, including 5 years of   history (2021-2025) and a `filingDate` field separate from the fiscal period-end date (e.g. period end "2025-09-27" vs filing date "2025-10-31") — solving the point-in-time/look-ahead bias issue that yfinance could not.

**Conclusion:** 
FMP's data structure is genuinely better than yfinance's — it includes real filing dates, which is critical for avoiding look-ahead bias. However, the free tier only covers US-listed companies. Since this project's universe is FTSE 350 (UK), FMP's free tier is not usable as-is. A paid FMP plan with international coverage would be required to use this source, or an institutional source (Capital IQ / LSEG Datastream via Warwick) is a better fit given it's free through the university.

**Next step:** Test Capital IQ (via Warwick access) on the same sample to check whether it resolves both remaining gaps — UK coverage and filing dates.

## Data Source Decision — yfinance (Phase 1)

**Why not Capital IQ Pro:** Capital IQ Pro was tested and found to have excellent data — full FTSE 350 coverage, deep historical range (RELX data visible back to the 1990s), and explicit filing dates on as-reported filings, solving the two biggest gaps found in yfinance and FMP. However, Excel plugin 
login access could not be resolved (authentication issue between web login and desktop plugin), blocking programmatic/bulk data extraction needed for a 350-company pipeline. Manual pulls for hundreds of companies are not feasible. Capital IQ remains a planned extension if access is resolved later.

**Decision:** Proceed with yfinance for the initial pipeline build.

**Known limitations and how they'll be handled:**
1. **~5 years of historical fundamentals only** (vs. desired 2013+ range). 
   Accepted for Phase 1 — backtest window will be scoped to available history. Extending to a longer window is a planned future step, either via Capital IQ (if access is fixed) or a paid data source.
2. **No filing/report dates** — only fiscal period-end dates available. 
   Addressed with a fixed reporting-lag assumption: fundamentals are treated as "known" 90 days after fiscal period-end, approximating UK reporting deadlines (interim results due within 3 months, annual results within 4 months of period-end). This is a documented simplification, not a perfect fix, and will be flagged as a limitation in the final memo.

**Design note:** the data pipeline will be built so the *source* is a swappable component — if a better data source becomes available later, only the data-fetching layer should need to change, not the downstream factor scoring or backtesting logic.